TombEngine/Game/effects/lightning.cpp

```diff
@@ -26,9 +29,168 @@ namespace TEN::Effects::Lightning
 	short LightningBuffer[1024];
-	
+		
 	std::vector<LIGHTNING_INFO> Lightning;
 
+	 TWOGUNINFO twogun[2];
+
+	 void TriggerLaserBeam(Vector3i src, Vector3i dest, short LeftRight)
+	 {
+		 TWOGUNINFO* tg;
+
+		 int j;
+		 for (int i = 0; i < 2; i++, i++)
+		 {
+			 tg = &twogun[LeftRight];
+			 for (j = 0; j < 4; j++)
+			 {
+				 if (tg->life == 0 || j == 3)
+					 break;
+			 }
+
+			 tg->pos1 = src;
+			 tg->pos4 = dest;
+			 tg->pos.Position = src;
+			 tg->life = 17;
+			 tg->spin = (GetRandomControl() & 31) << 11;
+			 tg->dlength = 4096;
+			 tg->r = 0;
+			 tg->b = 255;
+			 tg->g = 96;
+			 tg->fadein = 8;
+
+			 TriggerLightningGlow(tg->pos.Position.x, tg->pos.Position.y, tg->pos.Position.z, 64 + (GetRandomControl() & 3) << 24, 0, tg->g >> 1, tg->b >> 1);
+			 TriggerLightning(&src, &dest, 1, 0, tg->g, tg->b, 20, (LI_THININ | LI_THINOUT), 19, 5);	
+			 TriggerLightning(&src, &dest, 1, 110, 255, 250, 20, (LI_THININ | LI_THINOUT), 4, 5);
+		 }
+	 }
+
+	void UpdateTwogunLasers()
+	{
+		for (int i = 0; i < 2; i++)
+		{
+			TWOGUNINFO* tg = &twogun[i];
+			int	j;
+
+			for (j = 0; j < 4; j++, tg++)
+			{
+				if (tg->life)
+				{
+					tg->life--;
+
+					if (tg->life < 16)
+					{
+						tg->spinadd -= tg->spinadd / 8;
+						tg->size++;
+					}
+					else if (tg->life == 16)
+					{
+						tg->spinadd = MAX_VISIBILITY_DISTANCE;
+						tg->coil = MAX_VISIBILITY_DISTANCE;
+						tg->length = tg->dlength;
+						tg->size = 4;
+					}
+					else
+					{
+						tg->coil += (MAX_VISIBILITY_DISTANCE - tg->coil) / 8;
+						if (tg->dlength - tg->length <= (tg->dlength / 4))
+						{
+							tg->spinadd += (MAX_VISIBILITY_DISTANCE - tg->spinadd) / 8;
+							tg->length = tg->dlength;
+						}
+						else
+							tg->length += (tg->dlength - tg->length) >> 2;
+						if (tg->size < 4)
+							tg->size++;
+					}
+
+					if (tg->fadein < 8)
+						tg->fadein++;
+
+					tg->spin -= tg->spinadd;
+				}
+			}
+		}
+	}
+
+	void CurlSpline(Vector3i* pos, short* buffer, TWOGUNINFO* tg)
+	{
+		buffer[0] = pos[0].x;
+		buffer[1] = pos[0].y;
+		buffer[2] = pos[0].z;
+
+		buffer += 4;
+
+		int segments =  tg->segments ;
+
+		float x, y, z;
+
+		z = 0;
+
+		Vector3 lineStartPoint = Vector3(pos[0].x, pos[0].y, pos[0].z);
+		Vector3 lineEndPoint = Vector3(pos[1].x, pos[1].y, pos[1].z);
+
+		// Calculate the direction vector of the line
+		Vector3 direction = (lineEndPoint - lineStartPoint) / segments;
+
+		// Normalize the direction vector
+		direction.Normalize();
+
+		// Calculate the length of the line
+		float lineLength = (lineEndPoint - lineStartPoint).Length() ;
+
+		// Calculate the step size along the line
+		float stepSize = lineLength / segments;
+
+		// Set the starting position to the line start point
+		Vector3 currentPos = lineStartPoint;
+		
+		short angle = tg->spin ;
+
+		int x1;
+		short radius = 0;
+
+		if (1 * tg->segments  > 0)
+		{
+			for (int i =  tg->segments ; i > 0; i--)
+			{
+				x1 = radius >> 1;
+				if (x1 > 48)
```

Comment:
You can use `std::clamp(x1, 0, 48)` here instead, also, please demagic `48`.

Reply:
fixed